drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,426 @@
+            case A15:
+                cfg->channels[15].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 15;
+                break;
+
+            case C8:
+                cfg->channels[16].pfpamsk = 1;
+            case B0:
+                cfg->channels[16].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 16;
+                break;
+
+            case C9:
+                cfg->channels[17].pfpamsk = 1;
+            case B1:
+                cfg->channels[17].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 17;
+                break;
+
+            case C10:
+                cfg->channels[18].pfpamsk = 1;
+            case B2:
+                cfg->channels[18].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 18;
+                break;
+
+            case C11:
+                cfg->channels[19].pfpamsk = 1;
+            case B3:
+                cfg->channels[19].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 19;
+                break;
+
+            case C12:
+                cfg->channels[20].pfpamsk = 1;
+            case B4:
+                cfg->channels[20].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 20;
+                break;
+
+            case C13:
+                cfg->channels[21].pfpamsk = 1;
+            case B5:
+                cfg->channels[21].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 21;
+                break;
+
+            case C14:
+                cfg->channels[22].pfpamsk = 1;
+            case B6:
+                cfg->channels[22].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 22;
+                break;
+
+            case C15:
+                cfg->channels[23].pfpamsk = 1;
+            case B7:
+                cfg->channels[23].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 23;
+                break;
+        }
+    }
+}
+void rgb_callback(PWMDriver *pwmp);
+
+void shared_matrix_rgb_enable(void) {
+    pwmcfg.callback = rgb_callback;
+    pwmEnablePeriodicNotification(&PWMD1);
+}
+
+void shared_matrix_rgb_disable_pwm(void) {
+    // Disable PWM outputs on column pins
+    for(uint8_t y = 0; y < LED_MATRIX_COLS; y++) {
+        pwmDisableChannel(&PWMD1,chan_col_order[y]);
+    }
+}
+
+void shared_matrix_rgb_disable_leds(void) {
+    // Disable LED outputs on RGB channel pins
+    for (uint8_t x = 0; x < LED_MATRIX_ROWS_HW; x++) {
+        writePinLow(led_row_pins[x]);
+    }
+}
+
+void update_pwm_channels(PWMDriver *pwmp) {
+    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
+    for(uint8_t col_idx = 0; col_idx < LED_MATRIX_COLS; col_idx++, row_shifter <<= 1) {
+        #if(DIODE_DIRECTION == ROW2COL)
+            // Scan the key matrix column
+          matrix_read_rows_on_col(current_matrix,col_idx,row_shifter);
+        #endif
+        uint8_t led_index = g_led_config.matrix_co[row_idx][col_idx];
```

Comment:
How is this code supposed to work with LEDs that are not associated with any particular key (e.g., the underglow LEDs that are mentioned in some other parts of the code)?

One obvious option is to do `uint8_t led_index = row_idx * LED_MATRIX_COLS + col_idx` (with the appropriately sized `led_state[]`), but this would result in having to describe some nonexistent LEDs in `g_led_config`, which is probably not desired. A general solution which would support underglow and other non-key LEDs would probably require an additional mapping layer (similar to `g_is31_leds[]`).

Reply:
The highlighted part is not the problem — that's how underglow in rgb_matrix is done in most cases.  The issue is that those LEDs must then be added into `matrix_co` as if they were associated with some key, so the key matrix must have the appropriate number of unused locations.  Then with some hardware you might find that you need to do things like adding dummy rows or columns to the key matrix if the LED matrix happens to be larger than the key matrix (because otherwise you would not be able to add `matrix_co` entries for those LEDs).

In general, it could be possible to have a hardware design where the layouts of the LED matrix and key matrix are completely unrelated, and only some GPIO pins happen to be shared between them; but maybe there are no actual boards designed like that, and then your solution saves some memory by reusing the existing `matrix_co` data to describe the LED matrix layout.  And if some weird board comes around, it could be possible to extend this code by doing something like this:
```c
#ifdef SN32_CUSTOM_LED_MAPPING
        uint8_t led_index = sn32_led_mapping[row_idx][col_idx];
#else
        uint8_t led_index = g_led_config.matrix_co[row_idx][col_idx];
#endif
```
and
```c
#ifdef SN32_CUSTOM_LED_MAPPING
extern const uint8_t sn32_led_mapping[LED_MATRIX_ROWS][LED_MATRIX_COLS];
#endif
```
(in this case `LED_MATRIX_ROWS` and `LED_MATRIX_COLS` may be larger than `MATRIX_ROWS` and `MATRIX_COLS`, and any mismatch between the layout of key matrix and LED matrix may be handled too).